pennylane/templates/embeddings.py

```diff
@@ -295,6 +295,228 @@ def BasisEmbedding(features, wires):
+                weights = pars[0]
+                features = pars[1]
+                QAOAEmbedding(features=features, weights=weights, wires=range(2))
+                return qml.expval(qml.PauliZ(0))
+
+
+            features = [1., 2.]
+            weights = [[0.1, -0.3, 1.5], [3.1, 0.2, -2.8]]
+            pars = [weights, features]
+
+            o = GradientDescentOptimizer()
+            for i in range(10):
+                pars = o.step(circuit2, pars)
+                print("Step ", i, " weights = ", pars[0], " features = ", pars[1])
+
+        **Local Fields**
+
+        While by default, ``RY`` gates are used as local fields, one may also choose ``local_field='Z'`` or
+        ``local_field='X'`` as hyperparameters of the embedding.
+
+        .. code-block:: python
+
+            @qml.qnode(dev)
+            def circuit(weights, f=None):
+                QAOAEmbedding(features=f, weights=weights, wires=range(2), local_field='Z')
+                return qml.expval(qml.PauliZ(0))
+
+        Choosing ``'Z'`` fields implements a QAOAEmbedding where the second Hamiltonian is a
+        1-dimensional Ising model.
+
+    """
+    #############
+    # Input checks
+    wires, n_wires = _check_wires(wires)
+
+    n_features = _get_shape(features)[0]
+    msg = "QAOAEmbedding cannot process more features than number of qubits {};" \
+          "got {}.".format(n_wires, len(features))
+    _check_shape(features, (n_wires,), bound='max', msg=msg)
+
+    msg = "Option for local field not known. Has to be one of ``'X'``, ``'Y'``, or ``'Z'``."
+    _check_hyperp_is_in_options(local_field, ['X', 'Y', 'Z'], msg=msg)
+    if local_field == 'Z':
+        local_field = RZ
+    elif local_field == 'X':
+        local_field = RX
+    else:
+        local_field = RY
+
+    repeat = _check_number_of_layers([weights])
+
+    weights = np.array(weights)
+    weights_shape = weights.shape
+    if n_wires == 1:
+        msg = "QAOAEmbedding with 1 qubit and {} layers requires weight " \
+              "array of shape {}; got {}".format(repeat, (repeat, 1), weights_shape)
+        _check_shape(weights, (repeat, 1), msg=msg)
+    elif n_wires == 2:
+        msg = "QAOAEmbedding with 2 qubits and {} layers requires weight " \
+              "array of shape {}; got {}".format(repeat, (repeat, 3), weights_shape)
+        _check_shape(weights, (repeat, 3), msg=msg)
+    else:
+        msg = "QAOAEmbedding with {} qubits and {} layers requires weight " \
+              "array of shape {}; got {}".format(n_wires, repeat, (repeat, 2*n_wires), weights_shape)
+        _check_shape(weights, (repeat, 2*n_wires), msg=msg)
+    #####################
+
+    for l in range(repeat):
+
+        # encode inputs into RX gates
+        for i in range(n_wires):
+            # Either feed in feature
+            if i < n_features:
+                RX(features[i], wires=wires[i])
+            # or a Hadamard
+            else:
+                Hadamard(wires=wires[i])
```

Comment:
Perhaps it might be useful for readers to have each logically separable "part" of the layer represented by a separate helper function. For example, this section could be organized into ``encode_inputs_into_rx_gates()`` function (following the content of the comment). This way, the code is self-documented, and just by looking into the for loop, one can easily see the steps taken in each layer.

Reply:
That's not a bad idea at all, thanks!